### PR TITLE
haskell: Add arguments for run- and test-time dependencies.

### PR DESCRIPTION
**Pull request by teh**
###### Motivation for this change

This is motivated by https://github.com/NixOS/nixpkgs/issues/17226

I too have the issue that several of my haskell binaries run programs that are in PATH, i.e. not hard-coded.

This requires a rebuild of ghc 7.10 and 8.0, and potentially a mass-rebuild.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 1]
```diff
@@ -321,13 +336,11 @@ stdenv.mkDerivation ({
 // optionalAttrs (postBuild != "")      { inherit postBuild; }
 // optionalAttrs (doCheck)              { inherit doCheck; }
 // optionalAttrs (checkPhase != "")     { inherit checkPhase; }
-// optionalAttrs (preCheck != "")       { inherit preCheck; }
```
Comment (LnL7) [inline pkgs/development/haskell-modules/generic-builder.nix]: These lines are for overrides, don't remove them.

**Comment by peti** [comment]
I am unhappy with this approach for the following reasons:

- Wrapping an executable is going to fix that particular use case, but very often Haskell packages offer similar functionality in a library, too, which defines the actual functions that the executable merely provides a standardized front-end too. For users of the library, this solution is not going to work, so the solution is incomplete.

- `testtimeDepends` doesn't serve any purpose. Instead, the proper attribute `testToolDepends` should be used.

- Haskell builds are auto-generated, but our generator is never going to produce any of these fields because that information is not available from Cabal files. So the only place this kind of information could possibly be specified is in overrides -- which means that adding attributes to the generic Haskell builder is unnecessary. Instead, this should be implemented as an external combinator function that lives in, say, haskell.lib.

- There is nothing Haskell-specific about the generation of these wrapper scripts, so the generic Haskell builder is not the right place to host that functionality.

**Comment by teh** [comment]
Hm OK - I tried the combinators first but had some trouble making them pick up wrapProgram because that seems to be a setup hook. I'll close and try to find a different approach.

**Comment by peti** [comment]
Adding `makeWrappers` to `executableToolDepends` should do the trick, but I haven't tested it.

**Comment by peti** [comment]
Anyhow, this feature should really be integrated with to the *generic builder* since there is nothing Haskell-specific about it.

**Pull request closed without merging**